src/modules/Game.class.js

```diff
@@ -2,67 +2,276 @@
+      }
+    } else if (initialState && Array.isArray(initialState)) {
+      this.board = initialState;
+    }
+    this.score = 0;
+    this.status = 'idle';
+  }
+
+  moveLeft() {
+    if (this.status !== 'playing' && this.status !== 'win') {
+      return;
+    }
+
+    const oldBoard = JSON.parse(JSON.stringify(this.board));
+
+    for (let i = 0; i < 4; i++) {
+      this.board[i] = this.processRow(this.board[i]);
+    }
+
+    const hasChanged = this.board.some((row, i) => {
+      return row.some((cell, j) => cell !== oldBoard[i][j]);
+    });
+
+    if (hasChanged) {
+      this.allRandomTile();
+    }
+
+    const has2048 = this.board.some((row) => row.includes(2048));
+
+    if (has2048 && this.status !== 'win') {
+      this.status = 'win';
+    }
+
+    if (!this.canMove() && !has2048) {
+      this.status = 'lose';
+    }
   }
 
-  moveLeft() {}
-  moveRight() {}
-  moveUp() {}
-  moveDown() {}
-
-  /**
-   * @returns {number}
-   */
-  getScore() {}
-
-  /**
-   * @returns {number[][]}
-   */
-  getState() {}
-
-  /**
-   * Returns the current game status.
-   *
-   * @returns {string} One of: 'idle', 'playing', 'win', 'lose'
-   *
-   * `idle` - the game has not started yet (the initial state);
-   * `playing` - the game is in progress;
-   * `win` - the game is won;
-   * `lose` - the game is lost
-   */
-  getStatus() {}
-
-  /**
-   * Starts the game.
-   */
-  start() {}
-
-  /**
-   * Resets the game.
-   */
-  restart() {}
-
-  // Add your own methods here
+  moveRight() {
+    if (this.status !== 'playing' && this.status !== 'win') {
+      return;
+    }
+
+    const oldBoard = JSON.parse(JSON.stringify(this.board));
+
+    for (let i = 0; i < 4; i++) {
+      this.board[i] = this.processRow(this.board[i].reverse()).reverse();
+    }
+
+    const hasChanged = this.board.some((row, i) => {
+      return row.some((cell, j) => cell !== oldBoard[i][j]);
+    });
+
+    if (hasChanged) {
+      this.allRandomTile();
+    }
+
+    const has2048 = this.board.some((row) => row.includes(2048));
+
+    if (has2048 && this.status !== 'win') {
+      this.status = 'win';
+    }
+
+    if (!this.canMove() && !has2048) {
+      this.status = 'lose';
+    }
+  }
+
+  moveUp() {
+    if (this.status !== 'playing' && this.status !== 'win') {
+      return;
+    }
+
+    const oldBoard = JSON.parse(JSON.stringify(this.board));
+    const transposed = [];
+
+    for (let i = 0; i < 4; i++) {
+      transposed[i] = [];
+
+      for (let j = 0; j < 4; j++) {
+        transposed[i][j] = this.board[j][i];
+      }
+    }
+
+    for (let i = 0; i < 4; i++) {
+      transposed[i] = this.processRow(transposed[i]);
+    }
+
+    for (let i = 0; i < 4; i++) {
+      for (let j = 0; j < 4; j++) {
+        this.board[i][j] = transposed[j][i];
+      }
+    }
+
+    const hasChanged = this.board.some((row, i) => {
+      return row.some((cell, j) => cell !== oldBoard[i][j]);
+    });
+
+    if (hasChanged) {
+      this.allRandomTile();
```

Comment:
Similar logic is in every method, try to move it to a helper function and use it everywhere